security/pkg/nodeagent/caclient/providers/vault1/client_test.go

```diff
@@ -0,0 +1,350 @@
+	fakeCert        = []string{"fake-certificate", "fake-ca1", "fake-ca2"}
+	vaultNonTLSAddr = "http://35.247.45.173:8200"
+	vaultTLSAddr    = "https://35.233.249.249:8200"
+)
+
+type MockVaultServer struct {
+	httpServer     *httptest.Server
+	loginRole      string
+	token          string
+	vaultLoginResp string
+	vaultSignResp  string
+}
+
+type clientConfig struct {
+	tls              bool
+	tlsCert          []byte
+	vaultAddr        string
+	vaultLoginRole   string
+	vaultLoginPath   string
+	vaultSignCsrPath string
+	clientToken      string
+	csr              []byte
+}
+
+type loginRequest struct {
+	Jwt  string `json:"jwt"`
+	Role string `json:"role"`
+}
+
+type signRequest struct {
+	Format string `json:"format"`
+	Csr    string `json:"csr"`
+}
+
+func TestClientOnMockVaultCA(t *testing.T) {
+	testCases := map[string]struct {
+		cliConfig    clientConfig
+		expectedCert []string
+		expectedErr  string
+	}{
+		"Valid certs 1": {
+			cliConfig:    clientConfig{tls: false, tlsCert: []byte{}, vaultLoginPath: "login", vaultSignCsrPath: "sign", clientToken: "fake-client-token", csr: []byte{01}},
+			expectedCert: fakeCert,
+			expectedErr:  "",
+		},
+		"Valid certs 1 (TLS)": {
+			cliConfig:    clientConfig{tls: true, vaultLoginPath: "login", vaultSignCsrPath: "sign", clientToken: "fake-client-token", csr: []byte{01}},
+			expectedCert: fakeCert,
+			expectedErr:  "",
+		},
+		"Wrong Vault addr": {
+			cliConfig:    clientConfig{tls: false, tlsCert: []byte{}, vaultAddr: "wrong-vault-addr", vaultLoginPath: "login", vaultSignCsrPath: "wrong-sign-path", clientToken: "fake-client-token", csr: []byte{01}},
+			expectedCert: nil,
+			expectedErr:  "failed to login Vault",
+		},
+		"Wrong login path": {
+			cliConfig:    clientConfig{tls: false, tlsCert: []byte{}, vaultLoginPath: "wrong-login-path", vaultSignCsrPath: "sign", clientToken: "fake-client-token", csr: []byte{01}},
+			expectedCert: nil,
+			expectedErr:  "failed to login Vault",
+		},
+		"Wrong client token": {
+			cliConfig:    clientConfig{tls: false, tlsCert: []byte{}, vaultLoginPath: "login", vaultSignCsrPath: "sign", clientToken: "wrong-client-token", csr: []byte{01}},
+			expectedCert: nil,
+			expectedErr:  "failed to login Vault",
+		},
+		"Wrong sign path": {
+			cliConfig:    clientConfig{tls: false, tlsCert: []byte{}, vaultLoginPath: "login", vaultSignCsrPath: "wrong-sign-path", clientToken: "fake-client-token", csr: []byte{01}},
+			expectedCert: nil,
+			expectedErr:  "failed to sign CSR",
+		},
+	}
+
+	for id, tc := range testCases {
+		ch := make(chan *MockVaultServer)
+		go func() {
```

Comment:
In stead of creating the server each time, can you create two servers in the beginning, one with TLS and one without TLS?

Reply:
Done.